aten/src/ATen/native/cpu/SortingKernel.cpp

```diff
@@ -60,7 +60,8 @@ void _dim_apply(
         }
       };
 
-      iter.for_each(loop);
+      int64_t grain_size = internal::GRAIN_SIZE / std::max(int64_t{1}, dim_size);
+      iter.for_each(loop, /*grain_size=*/grain_size);
```

Comment:
`GRAIN_SIZE` is defined to be equal to `32768`, which means for `dim_size` larger than this value `grain_size` will be zero. This in turn implies that at least for the OpenMP implementation of the parallel for-loop `grain_size` is going to be completely ignored, and the true granularity is going to be `prod(size[:dim] + size[dim+1:]) / num_threads`. This quantity could be larger than `GRAIN_SIZE`. Let's benchmark shapes `(GRAIN_SIZE, 1024), (1024, GRAIN_SIZE), (GRAIN_SIZE, 64, 64), (64, GRAIN_SIZE, 64), (64, 64, GRAIN_SIZE)`.

Reply:
If that is more comfortable for you, you can use the pytorch's `benchmark` package. An example script could be taken from, for example, https://github.com/pytorch/pytorch/pull/72710#issue-1132586108. It is in the header.

Reply:
Oh, I understand your worries, do you think it's more reasonable to set grain_size to std::max((int64_t) 1, at::internal::GRAIN_SIZE / dim_size) ? and actually these two setting ways for grain_size both are commonly used in PyTorch.

Reply:
I don't see any problem here. `grain_size=0` should be supported by all `parallel_for` implementations, as meaning the maximum number of threads possible will be launched. This is exactly the behaviour you'd expect because each loop index will have more than `internal::GRAIN_SIZE` elements, so each is granular enough to warrant a thread.

`grain_size=0` is also used explicitly in some kernels, e.g.
https://github.com/pytorch/pytorch/blob/495aa9bc3adf88e1499a042323871dc4650c7c6f/aten/src/ATen/native/AdaptiveMaxPooling3d.cpp#L99

Reply:
The problem is not the value, the problem is that when it is zero, there are cases when smaller threads are going to get launcher for lager inputs.
Could you please do the benchmarks with all the suggested shapes and with all backends to claim that such a grain_size is indeed improving things uniformly no matter what the implementation of parallel-for is?

Reply:
This is not true that all threads will have more than `GRAIN_SIZE` elements.

Reply:
>I agree about grain_size=0, but once this value is different, we could expect divergent behaviors between different backends. 

The backends are only very slightly different, not enought to make a huge difference

>But then why wouldn't we just max-out from the get-go?

Because you will get worse performance on small tensors, because of the extra parallel overhead.

Reply:
@nikitaved, I benchmarked shapes (GRAIN_SIZE + 1, 1024), (1024, GRAIN_SIZE + 1), (GRAIN_SIZE + 1, 64, 64), (64, GRAIN_SIZE + 1, 64), (64, 64, GRAIN_SIZE + 1), (GRAIN_SIZE + 1, GRAIN_SIZE + 1) with OMP, NATIVE, TBB backend.
[sort_benchmark_the_pr_TBB.log](https://github.com/pytorch/pytorch/files/9099302/sort_benchmark_the_pr_TBB.log)
[sort_benchmark_the_pr_NATIVE.log](https://github.com/pytorch/pytorch/files/9099303/sort_benchmark_the_pr_NATIVE.log)
[sort_benchmark_the_pr_OMP.log](https://github.com/pytorch/pytorch/files/9099304/sort_benchmark_the_pr_OMP.log)
[sort_benchmark_08c32d7_TBB.log](https://github.com/pytorch/pytorch/files/9099306/sort_benchmark_08c32d7_TBB.log)
[sort_benchmark_08c32d7_NATIVE.log](https://github.com/pytorch/pytorch/files/9099307/sort_benchmark_08c32d7_NATIVE.log)
[sort_benchmark_08c32d7_OMP.log](https://github.com/pytorch/pytorch/files/9099308/sort_benchmark_08c32d7_OMP.log)